packages/crayons-core/src/utils/draggable.ts

```diff
@@ -0,0 +1,302 @@
+    dragElement = e.target;
+    this.cancelDebouncedDrag = false;
+
+    // Set dragElementId for Firefox
+    e.dataTransfer.setData('text/plain', dragElement.id);
+
+    // Set all items inside the drag container except the current element
+    const draggingElementIndex = this.childElements.indexOf(dragElement);
+    this.nextSibling = this.childElements[draggingElementIndex + 1];
+    this.childElements.splice(draggingElementIndex, 1);
+  };
+
+  private onDragEnter = (e) => {
+    if (!dragElement) {
+      return;
+    }
+    const sortContainer = e.composedPath().find((el) => el.id === this.host.id);
+    if (sortContainer && sortContainer !== this.previousContainer) {
+      // the drag element have entered or re-entered current container(this.host)
+      this.cancelDebouncedDrag = false;
+    }
+    this.previousContainer = sortContainer;
+  };
+
+  private onDragLeave = (e) => {
+    if (!dragElement) {
+      return;
+    }
+    const outTarget = e.fromElement || e.relatedTarget;
+    if (!e.currentTarget.contains(outTarget)) {
+      // Check whether the outTarget's host (in case of shadow dom) exists in currentTarget
+      const parentHost = this.getMatchingHost(
+        outTarget,
+        this.host.children[0].tagName
+      );
+      if (!e.currentTarget.contains(parentHost)) {
+        // the drag element have left the current container(this.host)
+        this.previousContainer = undefined;
+        this.cancelDebouncedDrag = true;
+
+        if (dragElement.parentElement.id !== this.host.id) {
+          // dragElement from another container
+          this.removePlaceholder();
+        } else {
+          // Sort within the same container
+          this.addElement(dragElement, this.nextSibling);
+        }
+      }
+    }
+  };
+
+  private onDragOver = (e) => {
+    e.preventDefault();
+    if (!dragElement) {
+      return;
+    }
+    const sortContainerId = dragElement.parentElement.id;
+    if (this.acceptFrom.includes(sortContainerId)) {
+      this.debouncedSetElement(this.childElements, dragElement, e.clientY);
+    }
+  };
+
+  // Both dragend and drop need to used as the drop will be fired only on the container on which the drag is dropped
+  // and no on the container where drag is originated.
+  private onDragEnd = (e) => {
+    this.resetData(e);
+  };
+
+  private onDrop = (e) => {
+    if (!dragElement) {
+      return;
+    }
+    const sortContainerId = dragElement.parentElement.id;
+    if (!this.acceptFrom.includes(sortContainerId)) {
+      return;
+    }
+    const newElement = this.placeholder || dragElement;
+    const droppedIndex = [...this.host.children].indexOf(newElement);
+    if (this.placeholder) {
+      if (this.options.addOnDrop) {
+        const clone = cloneNodeWithEvents(dragElement, true, true);
+        this.placeholder.replaceWith(clone);
+      } else {
+        this.removePlaceholder();
+      }
+    }
+    this.host.dispatchEvent(
+      new CustomEvent('fwDropBase', {
```

Comment:
fwDropBase let's name this clearly, i am not able to follow what is base

Reply:
`fwDrop` event will be emitted by the component consuming this object. To prevent the name ambiguity i have named this event as `fwDropBase` as this event is emitted by the base component.